django/contrib/admin/helpers.py

```diff
@@ -198,10 +198,16 @@ def contents(self):
                 if boolean:
                     result_repr = _boolean_icon(value)
                 else:
-                    result_repr = smart_text(value)
-                    if getattr(attr, "allow_tags", False):
-                        result_repr = mark_safe(result_repr)
+                    if hasattr(value, "__html__"):
+                        result_repr = value
                     else:
+                        result_repr = smart_text(value)
+                        if getattr(attr, "allow_tags", False):
+                            warnings.warn(
+                                "The allow_tags attribute is deprecated. Return a SafeString instance by "
+                                "using format_html, format_html_join or mark_safe instead.",
+                                RemovedInDjango20Warning, stacklevel=2)
+                            result_repr = mark_safe(value)
                         result_repr = linebreaksbr(result_repr)
```

Comment:
seems like this line should be in an else branch (i.e. it's currently not applied to things with allow_tags=True)